symengine/test_visitors.h

```diff
@@ -213,6 +213,40 @@ class RealVisitor : public BaseVisitor<RealVisitor>
 
     tribool apply(const Basic &b);
 };
+
+class PolynomialVisitor : public BaseVisitor<PolynomialVisitor>
+{
+private:
+    bool is_polynomial_ = true;
+    bool variables_allowed_ = true;
+    const set_basic &variables_;
+
+    void check_power(const Basic &base, const Basic &exp);
+
+public:
+    PolynomialVisitor(const set_basic &variables) : variables_(variables)
+    {
+    }
+    void bvisit(const Basic &x);
+    void bvisit(const Number &x){};
+    void bvisit(const Constant &x){};
+    void bvisit(const Symbol &x);
+    void bvisit(const Add &x);
+    void bvisit(const Mul &x);
+    void bvisit(const Pow &x);
+    void bvisit(const Set &x)
+    {
+        is_polynomial_ = false;
+    };
+    void bvisit(const Relational &x)
+    {
+        is_polynomial_ = false;
+    };
+
+    bool apply(const Basic &b);
+};
+
+bool is_polynomial(const Basic &b, const set_basic &variables = {});
```

Comment:
I don't understand what `variables` is doing. Can you explain with a comment?

Reply:
I added a documentation comment. The `variables` is analogous to `syms` in sympy. See also https://github.com/sympy/sympy/issues/21088 for an explanation.